diff --git a/lalsimulation/src/LALSimIMR.h b/lalsimulation/src/LALSimIMR.h
index d4129de..448d78d 100644
--- a/lalsimulation/src/LALSimIMR.h
+++ b/lalsimulation/src/LALSimIMR.h
@@ -106,13 +106,13 @@ SphHarmTimeSeries *XLALSimIMREOBNRv2Modes(const REAL8 phiRef, const REAL8 deltaT
 
 double XLALSimIMRSpinAlignedEOBPeakFrequency(REAL8 m1SI, REAL8 m2SI, const REAL8 spin1z, const REAL8 spin2z, UINT4 SpinAlignedEOBversion);
 int XLALSimIMRSpinAlignedEOBWaveform(REAL8TimeSeries **hplus, REAL8TimeSeries **hcross, const REAL8 phiC, REAL8 deltaT, const REAL8 m1SI, const REAL8 m2SI, const REAL8 fMin, const REAL8 r, const REAL8 inc, const REAL8 spin1z, const REAL8 spin2z, UINT4 SpinAlignedEOBversion, LALDict *LALparams);
-int XLALSimIMRSpinAlignedEOBWaveformAll(REAL8TimeSeries **hplus, REAL8TimeSeries **hcross, const REAL8 phiC, REAL8 deltaT, const REAL8 m1SI, const REAL8 m2SI, const REAL8 fMin, const REAL8 r, const REAL8 inc, const REAL8 spin1z, const REAL8 spin2z, UINT4 SpinAlignedEOBversion, const REAL8 lambda2Tidal1, const REAL8 lambda2Tidal2, const REAL8 omega02Tidal1, const REAL8 omega02Tidal2, const REAL8 lambda3Tidal1, const REAL8 lambda3Tidal2, const REAL8 omega03Tidal1, const REAL8 omega03Tidal2, const REAL8 quadparam1, const REAL8 quadparam2, REAL8Vector *nqcCoeffsInput, const INT4 nqcFlag, LALValue *ModeArray);
+int XLALSimIMRSpinAlignedEOBWaveformAll(REAL8TimeSeries **hplus, REAL8TimeSeries **hcross, const REAL8 phiC, REAL8 deltaT, const REAL8 m1SI, const REAL8 m2SI, const REAL8 fMin, const REAL8 r, const REAL8 inc, const REAL8 spin1z, const REAL8 spin2z, UINT4 SpinAlignedEOBversion, const REAL8 lambda2Tidal1, const REAL8 lambda2Tidal2, const REAL8 omega02Tidal1, const REAL8 omega02Tidal2, const REAL8 lambda3Tidal1, const REAL8 lambda3Tidal2, const REAL8 omega03Tidal1, const REAL8 omega03Tidal2, const REAL8 quadparam1, const REAL8 quadparam2, const COMPLEX16 qnmfreq220, const COMPLEX16 qnmfreq210, const COMPLEX16 qnmfreq330, const COMPLEX16 qnmfreq440, const COMPLEX16 qnmfreq550, REAL8Vector *nqcCoeffsInput, const INT4 nqcFlag, LALValue *ModeArray);
 int XLALSimIMRSpinAlignedEOBModes(SphHarmTimeSeries ** hlmmode,
   //SM
   REAL8Vector ** dynamics_out, /**<< OUTPUT, low-sampling dynamics */
   REAL8Vector ** dynamicsHi_out, /**<< OUTPUT, high-sampling dynamics */
   //SM
- REAL8 deltaT, const REAL8 m1SI, const REAL8 m2SI, const REAL8 fMin, const REAL8 r, const REAL8 spin1z, const REAL8 spin2z, UINT4 SpinAlignedEOBversion, const REAL8 lambda2Tidal1, const REAL8 lambda2Tidal2, const REAL8 omega02Tidal1, const REAL8 omega02Tidal2, const REAL8 lambda3Tidal1, const REAL8 lambda3Tidal2, const REAL8 omega03Tidal1, const REAL8 omega03Tidal2, const REAL8 quadparam1, const REAL8 quadparam2, REAL8Vector *nqcCoeffsInput, const INT4 nqcFlag);
+ REAL8 deltaT, const REAL8 m1SI, const REAL8 m2SI, const REAL8 fMin, const REAL8 r, const REAL8 spin1z, const REAL8 spin2z, UINT4 SpinAlignedEOBversion, const REAL8 lambda2Tidal1, const REAL8 lambda2Tidal2, const REAL8 omega02Tidal1, const REAL8 omega02Tidal2, const REAL8 lambda3Tidal1, const REAL8 lambda3Tidal2, const REAL8 omega03Tidal1, const REAL8 omega03Tidal2, const REAL8 quadparam1, const REAL8 quadparam2, const COMPLEX16 qnmfreq220, const COMPLEX16 qnmfreq210, const COMPLEX16 qnmfreq330, const COMPLEX16 qnmfreq440, const COMPLEX16 qnmfreq550, REAL8Vector *nqcCoeffsInput, const INT4 nqcFlag);
 /*int XLALSimIMRSpinEOBWaveform(REAL8TimeSeries **hplus, REAL8TimeSeries **hcross, const REAL8 phiC, const REAL8 deltaT, const REAL8 m1SI, const REAL8 m2SI, const REAL8 fMin, const REAL8 r, const REAL8 inc, const REAL8 spin1[], const REAL8 spin2[]);
  */
 
diff --git a/lalsimulation/src/LALSimIMREOBHybridRingdown.c b/lalsimulation/src/LALSimIMREOBHybridRingdown.c
index 1b514f6..1c4719f 100644
--- a/lalsimulation/src/LALSimIMREOBHybridRingdown.c
+++ b/lalsimulation/src/LALSimIMREOBHybridRingdown.c
@@ -1025,6 +1025,11 @@ static UNUSED INT4 XLALSimIMREOBAttachFitRingdown(
     const REAL8 spin2x,    /**<<The spin of the second object; */
     const REAL8 spin2y,    /**<<The spin of the second object; */
     const REAL8 spin2z,    /**<<The spin of the second object; */
+    const COMPLEX16 qnmfreq220,
+    const COMPLEX16 qnmfreq210,
+    const COMPLEX16 qnmfreq330,
+    const COMPLEX16 qnmfreq440,
+    const COMPLEX16 qnmfreq550,
     REAL8Vector * timeVec, /**<< Vector containing the time values */
     REAL8Vector * matchrange,
                            /**<< Time values chosen as points for performing comb matching */
@@ -1083,9 +1088,24 @@ static UNUSED INT4 XLALSimIMREOBAttachFitRingdown(
     /*********************************************************************************************/
     /* Getting  QNMs */
     modefreqs = XLALCreateCOMPLEX16Vector(1);
-    if (XLALSimIMREOBGenerateQNMFreqV2(modefreqs, mass1, mass2, spin1, spin2, l, m, 1, appr) == XLAL_FAILURE) {
-        XLALDestroyCOMPLEX16Vector(modefreqs);
-        XLAL_ERROR(XLAL_EFUNC);
+    //if (XLALSimIMREOBGenerateQNMFreqV2(modefreqs, mass1, mass2, spin1, spin2, l, m, 1, appr) == XLAL_FAILURE) {
+    //    XLALDestroyCOMPLEX16Vector(modefreqs);
+    //    XLAL_ERROR(XLAL_EFUNC);
+    //}
+    if (( l == 2 ) && (m == 2)){
+	modefreqs->data[0] = qnmfreq220;
+    }
+    if (( l == 2 ) && ( m == 1)) {
+        modefreqs->data[0] = qnmfreq210;
+    }
+    if (( l == 3 ) && ( m == 3)) {
+        modefreqs->data[0] = qnmfreq330;
+    }
+    if (( l == 4 ) && ( m == 4)) {
+        modefreqs->data[0] = qnmfreq440;
+    }
+    if (( l == 5 ) && ( m == 5)) {
+        modefreqs->data[0] = qnmfreq550;
     }
 
     //RC: we use this variable to compute the damping time of the 22 mode which will be used to set the lenght of the ringdown for all the modes
diff --git a/lalsimulation/src/LALSimIMRSpinAlignedEOB.c b/lalsimulation/src/LALSimIMRSpinAlignedEOB.c
index 6c1a54e..e51de04 100644
--- a/lalsimulation/src/LALSimIMRSpinAlignedEOB.c
+++ b/lalsimulation/src/LALSimIMRSpinAlignedEOB.c
@@ -544,6 +544,18 @@ XLALSimIMRSpinAlignedEOBWaveform (REAL8TimeSeries ** hplus,	     /**<< OUTPUT, +
   REAL8 omega03Tidal2 = 0;
   REAL8 quadparam1 = 0;
   REAL8 quadparam2 = 0;
+  COMPLEX16 qnmfreq220 = 0;
+  COMPLEX16 qnmfreq210 = 0;
+  COMPLEX16 qnmfreq330 = 0;
+  COMPLEX16 qnmfreq440 = 0;
+  COMPLEX16 qnmfreq550 = 0;
+  
+
+  qnmfreq220 = XLALSimInspiralWaveformParamsLookupQNMFreq220(LALParams);
+  qnmfreq210 = XLALSimInspiralWaveformParamsLookupQNMFreq210(LALParams);
+  qnmfreq330 = XLALSimInspiralWaveformParamsLookupQNMFreq330(LALParams);
+  qnmfreq440 = XLALSimInspiralWaveformParamsLookupQNMFreq440(LALParams);
+  qnmfreq550 = XLALSimInspiralWaveformParamsLookupQNMFreq550(LALParams);
 
   lambda2Tidal1 = XLALSimInspiralWaveformParamsLookupTidalLambda1(LALParams);
   lambda2Tidal2 = XLALSimInspiralWaveformParamsLookupTidalLambda2(LALParams);
@@ -607,7 +619,7 @@ XLALSimIMRSpinAlignedEOBWaveform (REAL8TimeSeries ** hplus,	     /**<< OUTPUT, +
       printf("First run SEOBNRv4 to compute NQCs\n");
 #endif
       ret = XLALSimIMRSpinAlignedEOBWaveformAll (hplus, hcross, phiC, 1./32768, m1BH, m2BH, 2*pow(10.,-1.5)/(2.*LAL_PI)/((m1BH + m2BH)*LAL_MTSUN_SI/LAL_MSUN_SI), r, inc, spin1z, spin2z, 400,
-					 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, nqcCoeffsInput, nqcFlag, ModeArray);
+					 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, qnmfreq220, qnmfreq210, qnmfreq330, qnmfreq440, qnmfreq550, nqcCoeffsInput, nqcFlag, ModeArray);
       if (ret == XLAL_FAILURE){
         if ( nqcCoeffsInput ) XLALDestroyREAL8Vector( nqcCoeffsInput );
         if(ModeArray) XLALDestroyValue(ModeArray);
@@ -627,7 +639,7 @@ XLALSimIMRSpinAlignedEOBWaveform (REAL8TimeSeries ** hplus,	     /**<< OUTPUT, +
                                                  omega02Tidal1, omega02Tidal2,
                                                  lambda3Tidal1, lambda3Tidal2,
                                                  omega03Tidal1, omega03Tidal2,
-                                                 quadparam1, quadparam2,
+                                                 quadparam1, quadparam2, qnmfreq220, qnmfreq210, qnmfreq330, qnmfreq440, qnmfreq550,
                                                  nqcCoeffsInput, nqcFlag, ModeArray);
      if (ret == XLAL_FAILURE){
        if ( nqcCoeffsInput ) XLALDestroyREAL8Vector( nqcCoeffsInput );
@@ -710,6 +722,11 @@ XLALSimIMRSpinAlignedEOBModes (SphHarmTimeSeries ** hlmmode,
                      /**<< parameter kappa_1 of the spin-induced quadrupole for body 1, quadrupole is Q_A = -kappa_A m_A^3 chi_A^2 */
 				     const REAL8 quadparam2,
                      /**<< parameter kappa_2 of the spin-induced quadrupole for body 2, quadrupole is Q_A = -kappa_A m_A^3 chi_A^2 */
+				     const COMPLEX16 qnmfreq220,
+				     const COMPLEX16 qnmfreq210,
+				     const COMPLEX16 qnmfreq330,
+				     const COMPLEX16 qnmfreq440,
+				     const COMPLEX16 qnmfreq550,
                      REAL8Vector *nqcCoeffsInput,
                      /**<< Input NQC coeffs */
                      const INT4 nqcFlag
@@ -2469,7 +2486,6 @@ for ( UINT4 k = 0; k<nModes; k++) {
     printf("tattach = %.16f\n", rdMatchPoint->data[1]);
 #endif
 UINT4 indAmpMax = 0; //RC this variable is only used for SEOBNRv4HM. It stores the index of the attaching point of the 22 mode
-
 for ( UINT4 k = 0; k<nModes; k++) {
 
     modeL  = lmModes[k][0];
@@ -2565,7 +2581,7 @@ for ( UINT4 k = 0; k<nModes; k++) {
             if (XLALSimIMREOBAttachFitRingdown (sigReHi, sigImHi, modeL, modeM,
 					  deltaTHigh, m1, m2, spin1[0],
 					  spin1[1], spin1[2], spin2[0],
-					  spin2[1], spin2[2], &timeHi,
+					  spin2[1], spin2[2], qnmfreq220, qnmfreq210, qnmfreq330, qnmfreq440, qnmfreq550, &timeHi,
 					  rdMatchPoint,
 					  SpinAlignedEOBapproximant, &indAmpMax) ==
                 XLAL_FAILURE)
@@ -3236,6 +3252,11 @@ XLALSimIMRSpinAlignedEOBWaveformAll (REAL8TimeSeries ** hplus,
                      /**<< parameter kappa_1 of the spin-induced quadrupole for body 1, quadrupole is Q_A = -kappa_A m_A^3 chi_A^2 */
 				     const REAL8 quadparam2,
                      /**<< parameter kappa_2 of the spin-induced quadrupole for body 2, quadrupole is Q_A = -kappa_A m_A^3 chi_A^2 */
+				     const COMPLEX16 qnmfreq220,
+				     const COMPLEX16 qnmfreq210,
+				     const COMPLEX16 qnmfreq330,
+				     const COMPLEX16 qnmfreq440,
+				     const COMPLEX16 qnmfreq550,
                      REAL8Vector *nqcCoeffsInput,
                      /**<< Input NQC coeffs */
                      const INT4 nqcFlag,
@@ -3264,7 +3285,7 @@ XLALSimIMRSpinAlignedEOBWaveformAll (REAL8TimeSeries ** hplus,
                                                omega02Tidal1, omega02Tidal2,
                                                lambda3Tidal1, lambda3Tidal2,
                                                omega03Tidal1, omega03Tidal2,
-                                               quadparam1, quadparam2,
+                                               quadparam1, quadparam2, qnmfreq220, qnmfreq210, qnmfreq330, qnmfreq440, qnmfreq550,
                                                nqcCoeffsInput, nqcFlag) == XLAL_FAILURE){
                                                  if(dynamics) XLALDestroyREAL8Vector(dynamics);
                                                  if(dynamicsHi) XLALDestroyREAL8Vector(dynamicsHi);
diff --git a/lalsimulation/src/LALSimInspiral.c b/lalsimulation/src/LALSimInspiral.c
index 3c6e34d..d48ee08 100644
--- a/lalsimulation/src/LALSimInspiral.c
+++ b/lalsimulation/src/LALSimInspiral.c
@@ -344,6 +344,11 @@ int XLALSimInspiralChooseTDWaveform(
     REAL8 quadparam2 = 1.+XLALSimInspiralWaveformParamsLookupdQuadMon2(LALparams);
     REAL8 lambda1 = XLALSimInspiralWaveformParamsLookupTidalLambda1(LALparams);
     REAL8 lambda2 = XLALSimInspiralWaveformParamsLookupTidalLambda2(LALparams);
+    UNUSED COMPLEX16 qnmfreq220 = XLALSimInspiralWaveformParamsLookupQNMFreq220(LALparams);
+    UNUSED COMPLEX16 qnmfreq210 = XLALSimInspiralWaveformParamsLookupQNMFreq210(LALparams);
+    UNUSED COMPLEX16 qnmfreq330 = XLALSimInspiralWaveformParamsLookupQNMFreq330(LALparams);
+    UNUSED COMPLEX16 qnmfreq440 = XLALSimInspiralWaveformParamsLookupQNMFreq440(LALparams);
+    UNUSED COMPLEX16 qnmfreq550 = XLALSimInspiralWaveformParamsLookupQNMFreq550(LALparams);
     int amplitudeO = XLALSimInspiralWaveformParamsLookupPNAmplitudeOrder(LALparams);
     int phaseO =XLALSimInspiralWaveformParamsLookupPNPhaseOrder(LALparams);
 		/* Tidal parameters to be computed, if required, by universal relations */
diff --git a/lalsimulation/src/LALSimInspiralWaveformParams.c b/lalsimulation/src/LALSimInspiralWaveformParams.c
index e707ffa..182a39a 100644
--- a/lalsimulation/src/LALSimInspiralWaveformParams.c
+++ b/lalsimulation/src/LALSimInspiralWaveformParams.c
@@ -67,6 +67,11 @@ DEFINE_INSERT_FUNC(PNEccentricityOrder, INT4, "eccO", -1)
 DEFINE_INSERT_FUNC(PNSpinOrder, INT4, "spinO", -1)
 DEFINE_INSERT_FUNC(PNTidalOrder, INT4, "tideO", -1)
 
+DEFINE_INSERT_FUNC(QNMFreq220, COMPLEX16, "qnmfreq220", 0)
+DEFINE_INSERT_FUNC(QNMFreq210, COMPLEX16, "qnmfreq210", 0)
+DEFINE_INSERT_FUNC(QNMFreq330, COMPLEX16, "qnmfreq330", 0)
+DEFINE_INSERT_FUNC(QNMFreq440, COMPLEX16, "qnmfreq440", 0)
+DEFINE_INSERT_FUNC(QNMFreq550, COMPLEX16, "qnmfreq550", 0)
 
 DEFINE_INSERT_FUNC(TidalLambda1, REAL8, "lambda1", 0)
 DEFINE_INSERT_FUNC(TidalLambda2, REAL8, "lambda2", 0)
@@ -167,6 +172,12 @@ DEFINE_LOOKUP_FUNC(PNEccentricityOrder, INT4, "eccO", -1)
 DEFINE_LOOKUP_FUNC(PNSpinOrder, INT4, "spinO", -1)
 DEFINE_LOOKUP_FUNC(PNTidalOrder, INT4, "tideO", -1)
 
+DEFINE_LOOKUP_FUNC(QNMFreq220, COMPLEX16, "qnmfreq220", 0)
+DEFINE_LOOKUP_FUNC(QNMFreq210, COMPLEX16, "qnmfreq210", 0)
+DEFINE_LOOKUP_FUNC(QNMFreq330, COMPLEX16, "qnmfreq330", 0)
+DEFINE_LOOKUP_FUNC(QNMFreq440, COMPLEX16, "qnmfreq440", 0)
+DEFINE_LOOKUP_FUNC(QNMFreq550, COMPLEX16, "qnmfreq550", 0)
+
 DEFINE_LOOKUP_FUNC(TidalLambda1, REAL8, "lambda1", 0)
 DEFINE_LOOKUP_FUNC(TidalLambda2, REAL8, "lambda2", 0)
 DEFINE_LOOKUP_FUNC(TidalOctupolarLambda1, REAL8, "TidalOctupolarLambda1", 0)
@@ -259,6 +270,12 @@ DEFINE_ISDEFAULT_FUNC(PNEccentricityOrder, INT4, "eccO", -1)
 DEFINE_ISDEFAULT_FUNC(PNSpinOrder, INT4, "spinO", -1)
 DEFINE_ISDEFAULT_FUNC(PNTidalOrder, INT4, "tideO", -1)
 
+DEFINE_ISDEFAULT_FUNC(QNMFreq220, COMPLEX16, "qnmfreq220", 0)
+DEFINE_ISDEFAULT_FUNC(QNMFreq210, COMPLEX16, "qnmfreq210", 0)
+DEFINE_ISDEFAULT_FUNC(QNMFreq330, COMPLEX16, "qnmfreq330", 0)
+DEFINE_ISDEFAULT_FUNC(QNMFreq440, COMPLEX16, "qnmfreq440", 0)
+DEFINE_ISDEFAULT_FUNC(QNMFreq550, COMPLEX16, "qnmfreq550", 0)
+
 DEFINE_ISDEFAULT_FUNC(TidalLambda1, REAL8, "lambda1", 0)
 DEFINE_ISDEFAULT_FUNC(TidalLambda2, REAL8, "lambda2", 0)
 DEFINE_ISDEFAULT_FUNC(TidalOctupolarLambda1, REAL8, "TidalOctupolarLambda1", 0)
diff --git a/lalsimulation/src/LALSimInspiralWaveformParams.h b/lalsimulation/src/LALSimInspiralWaveformParams.h
index c8ebdaf..705605f 100644
--- a/lalsimulation/src/LALSimInspiralWaveformParams.h
+++ b/lalsimulation/src/LALSimInspiralWaveformParams.h
@@ -24,6 +24,11 @@ int XLALSimInspiralWaveformParamsInsertPNEccentricityOrder(LALDict *params, INT4
 int XLALSimInspiralWaveformParamsInsertPNSpinOrder(LALDict *params, INT4 value);
 int XLALSimInspiralWaveformParamsInsertPNTidalOrder(LALDict *params, INT4 value);
 
+int XLALSimInspiralWaveformParamsInsertQNMFreq220(LALDict *params, COMPLEX16 value);
+int XLALSimInspiralWaveformParamsInsertQNMFreq210(LALDict *params, COMPLEX16 value);
+int XLALSimInspiralWaveformParamsInsertQNMFreq330(LALDict *params, COMPLEX16 value);
+int XLALSimInspiralWaveformParamsInsertQNMFreq440(LALDict *params, COMPLEX16 value);
+int XLALSimInspiralWaveformParamsInsertQNMFreq550(LALDict *params, COMPLEX16 value);
 int XLALSimInspiralWaveformParamsInsertTidalLambda1(LALDict *params, REAL8 value);
 int XLALSimInspiralWaveformParamsInsertTidalLambda2(LALDict *params, REAL8 value);
 int XLALSimInspiralWaveformParamsInsertTidalOctupolarLambda1(LALDict *params, REAL8 value);
@@ -109,6 +114,11 @@ INT4 XLALSimInspiralWaveformParamsLookupPNEccentricityOrder(LALDict *params);
 INT4 XLALSimInspiralWaveformParamsLookupPNSpinOrder(LALDict *params);
 INT4 XLALSimInspiralWaveformParamsLookupPNTidalOrder(LALDict *params);
 
+COMPLEX16 XLALSimInspiralWaveformParamsLookupQNMFreq220(LALDict *params);
+COMPLEX16 XLALSimInspiralWaveformParamsLookupQNMFreq210(LALDict *params);
+COMPLEX16 XLALSimInspiralWaveformParamsLookupQNMFreq330(LALDict *params);
+COMPLEX16 XLALSimInspiralWaveformParamsLookupQNMFreq440(LALDict *params);
+COMPLEX16 XLALSimInspiralWaveformParamsLookupQNMFreq550(LALDict *params);
 REAL8 XLALSimInspiralWaveformParamsLookupTidalLambda1(LALDict *params);
 REAL8 XLALSimInspiralWaveformParamsLookupTidalLambda2(LALDict *params);
 REAL8 XLALSimInspiralWaveformParamsLookupTidalOctupolarLambda1(LALDict *params);
@@ -194,6 +204,11 @@ int XLALSimInspiralWaveformParamsPNEccentricityOrderIsDefault(LALDict *params);
 int XLALSimInspiralWaveformParamsPNSpinOrderIsDefault(LALDict *params);
 int XLALSimInspiralWaveformParamsPNTidalOrderIsDefault(LALDict *params);
 
+int XLALSimInspiralWaveformParamsQNMFreq220IsDefault(LALDict *params);
+int XLALSimInspiralWaveformParamsQNMFreq210IsDefault(LALDict *params);
+int XLALSimInspiralWaveformParamsQNMFreq330IsDefault(LALDict *params);
+int XLALSimInspiralWaveformParamsQNMFreq440IsDefault(LALDict *params);
+int XLALSimInspiralWaveformParamsQNMFreq550IsDefault(LALDict *params);
 int XLALSimInspiralWaveformParamsTidalLambda1IsDefault(LALDict *params);
 int XLALSimInspiralWaveformParamsTidalLambda2IsDefault(LALDict *params);
 int XLALSimInspiralWaveformParamsTidalOctupolarLambda1IsDefault(LALDict *params);
